cpp/dolfinx/common/IndexMap.cpp

```diff
@@ -32,6 +32,356 @@ std::array<std::vector<int>, 2> build_src_dest(MPI_Comm comm,
 
   return {std::move(src), std::move(dest)};
 }
+
+/// Given an index map and a subset of local indices (can be owned or ghost but
+/// must be unique and sorted), computes the owned, ghost and ghost owners in
+/// the submap.
+/// @param[in] imap An index map.
+/// @param[in] indices List of entity indices (indices local to the process).
+/// @pre `indices` must be sorted and unique.
+/// @return The owned, ghost, and ghost owners in the submap. All indices are
+/// local and with respect to the original index map.
+std::tuple<std::vector<std::int32_t>, std::vector<std::int32_t>,
+           std::vector<std::int32_t>>
+compute_submap_indices(const dolfinx::common::IndexMap& imap,
+                       std::span<const std::int32_t> indices)
+{
+  // --- Step 1 ---: Send ghost indices in `indices` to their owners
```

Comment:
For long functions with 'steps' it's useful to scope each step (by using a `{`, `}` pair).